File: haskell/secret-handshake/src/SecretHandshake.hs
{-# LANGUAGE FlexibleInstances #-}
module SecretHandshake (handshake) where

{-
Inspired by ed359's solution
-}


import qualified Data.Bits as Bits

bitFunctions :: [(Int, [String] -> [String])]
bitFunctions = 
    [ ( 0, (++ ["wink"]) )
    , ( 1, (++ ["double blink"]) )
    , ( 2, (++ ["close your eyes"]) )
    , ( 3, (++ ["jump"]) )
    , ( 4, reverse )
    ]


class HandshakeCode a where
    testBit :: a -> Int -> Bool

    handshake :: a -> [String]
    handshake code =
        let
            helper acc (bitIndex, action) =
                if testBit code bitIndex then
                    action acc
                else
                    acc
        in
            foldl helper [] bitFunctions


instance HandshakeCode Int where
    testBit = Bits.testBit


instance HandshakeCode String where
    testBit s bitIndex =
        let
            len = length s
            strIndex = len-bitIndex-1
        in
            if bitIndex >= len then
                False
            else
                (s !! strIndex) == '1'

File: go/secret-handshake/secret_handshake.go
package secret

const testVersion = 1

// Handshake output secret handshake for a given binary code
func Handshake(binary uint) []string {
	var handshake []string

	if binary&1 > 0 {
		handshake = append(handshake, "wink")
	}
	if binary&2 > 0 {
		handshake = append(handshake, "double blink")
	}
	if binary&4 > 0 {
		handshake = append(handshake, "close your eyes")
	}
	if binary&8 > 0 {
		handshake = append(handshake, "jump")
	}
	if binary&16 == 0 {
		return handshake
	}
	var rev []string
	var n = len(handshake)
	for i := n - 1; i >= 0; i-- {
		rev = append(rev, handshake[i])
	}
	return rev
}
